NPM generator: package names error

There's an error when **generating the npm graph,** you'll find this error at the end of the message.

The problem is that this "https://raw.githubusercontent.com/bconnorwhite/all-package-names/master/data/all.json?raw=true" URL returns "404: Not Found".
In fact, the repository owner of "all-package-names" has **removed the json file** and **replaced it with a .gz archive** to reduce its size.

I hotfixed this by replacing the URL line 16 of the "**NpmIdGenerator**" class with "https://raw.githubusercontent.com/bconnorwhite/all-package-names/2f4ce16fd68355fe366eab0abf37e97bd7cdd657/data/all.json?raw=true" (id of the last commit with the json file) and it seems to work.

The problem now is that we can't generate graphs more recent than this commit (Mar 31, 2023).

Thanks for your tool.

Error output:

> java -jar ./target/dgmf.jar start
> 12:48:52 [INFO ] [CommandUtilities] - Loaded configuration from system.properties
> 12:48:52 [INFO ] [StartCommand] - Initialise task
> 12:48:52 [INFO ] [StartCommand] - Started task with repository npm
> 12:48:53 [ERROR] [NpmIdGenerator] - Malformed GitHub response, cannot load package names
> 12:48:53 [ERROR] [Task] - => Mining of PackageIds for Repository npm failed <=
> 12:48:53 [ERROR] [ConsoleApplication] - Execution aborted due to errors
> 12:48:53 [INFO ] [ExceptionLogger] - Exception logs are empty

--- a/README.md
+++ b/README.md
@@ -28,19 +28,20 @@ DGMF outputs dependency graphs to a *Neo4j* graph database instance. Therefore, 
 
 To configure DGMF, you may edit all properties specified in the `system.properties` file. Alternatively, you can configure and control the DGMF interactively in the console when the `interactive-shell` flag is set. The following configuration properties are available:
 
-|Property | Values | Default Value | Description |
-|:---|:---|:---:|:---|
-|`dgm.repo` | {`pypi`, `maven`, `npm`, `nuget`} | npm | Select Repository to work on|
-|`dgm.parallel` | Positive Integers | 10 | Number of parallel streaming pipelines to use. Empirically, we found values between 10 and 30 to be a good fit for most architectures.|
-|`dgm.linkage` | {`pp`, `ap`, `aa`} | pp | Dependency resolution level to use for resolving dependency edges. Either Package-to-Package (`pp`), Artifact-to-Package (`ap`) or Artifact-to-Artifact (`aa`).|
-|`dgm.databaseaddress`| Strings | `bolt://localhost:7687` | Neo4j *Bolt Protocol* URL of form `bolt://<host>:<port>`|
-|`dgm.databaseusername`| Strings | `neo4j` | Username for connecting to Neo4j |
-|`dgm.databasepassword`| Strings | `neo4j` | Password for connecting to Neo4j |
-|`dgm.limit` | Positive Integers | 0 | If limit != 0, mining is stopped after processing the specified number of packages.|
-|`dgm.offset` | Positive Integers | 0 | If offset != 0, the specified number of packages are skipped when building a dependency graph.|
-|`dgm.interactive-shell` | {`true`, `false`} | `false` | If true, DGMF starts an interactive shell session. |
-|`dgm.import-ids`| {`true`, `false`} | `false` | If true, package ids are not generated live, but imported from an id file that was previously exported using DGMF.|
-|`dgm.id-file`| String | `<dgm.repo>_ids.txt`| Only applies if `dgm.import-ids` is `true`. Specifies path to file that holds package ids.|
+| Property                   | Values                            |      Default Value      | Description                                                                                                                                                         |
+|:---------------------------|:----------------------------------|:-----------------------:|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `dgm.repo`                 | {`pypi`, `maven`, `npm`, `nuget`} |           npm           | Select Repository to work on                                                                                                                                        |
+| `dgm.parallel`             | Positive Integers                 |           10            | Number of parallel streaming pipelines to use. Empirically, we found values between 10 and 30 to be a good fit for most architectures.                              |
+| `dgm.linkage`              | {`pp`, `ap`, `aa`}                |           pp            | Dependency resolution level to use for resolving dependency edges. Either Package-to-Package (`pp`), Artifact-to-Package (`ap`) or Artifact-to-Artifact (`aa`).     |
+| `dgm.databaseaddress`      | Strings                           | `bolt://localhost:7687` | Neo4j *Bolt Protocol* URL of form `bolt://<host>:<port>`                                                                                                            |
+| `dgm.databaseusername`     | Strings                           |         `neo4j`         | Username for connecting to Neo4j                                                                                                                                    |
+| `dgm.databasepassword`     | Strings                           |         `neo4j`         | Password for connecting to Neo4j                                                                                                                                    |
+| `dgm.limit`                | Positive Integers                 |            0            | If limit != 0, mining is stopped after processing the specified number of packages.                                                                                 |
+| `dgm.offset`               | Positive Integers                 |            0            | If offset != 0, the specified number of packages are skipped when building a dependency graph.                                                                      |
+| `dgm.interactive-shell`    | {`true`, `false`}                 |         `false`         | If true, DGMF starts an interactive shell session.                                                                                                                  |
+| `dgm.import-ids`           | {`true`, `false`}                 |         `false`         | If true, package ids are not generated live, but imported from an id file that was previously exported using DGMF.                                                  |
+| `dgm.id-file`              | String                            |  `<dgm.repo>_ids.txt`   | Only applies if `dgm.import-ids` is `true`. Specifies path to file that holds package ids.                                                                          |
+| `dgm.npm.commit-qualifier` | String                            |        `master`         | Sets which commit or branch of [Connor White's NPM package list](https://github.com/bconnorwhite/all-package-names) shall be used to generate the NPM package list. |
 
 ### Building DGMF locally
 You can build the DGMF executable `.jar` file locally on your machine. To do this, you need to execute to following command:

--- a/src/main/java/Application/Commands/StartCommand.java
+++ b/src/main/java/Application/Commands/StartCommand.java
@@ -46,6 +46,6 @@ public class StartCommand implements Command {
 
     @Override
     public String getDescription() {
-        return "- Command 'start <repository>(optional)': starts a new task with given repository";
+        return "- Command 'start [<repository>]': starts a new task with given repository";
     }
 }

--- a/src/main/java/Application/Commands/UpdateCommand.java
+++ b/src/main/java/Application/Commands/UpdateCommand.java
@@ -44,6 +44,6 @@ public class UpdateCommand implements Command {
 
     @Override
     public String getDescription() {
-        return null;
+        return "- Command 'update [<repository>]': update the graph of the given repository";
     }
 }

--- a/src/main/java/Application/ConsoleApplication.java
+++ b/src/main/java/Application/ConsoleApplication.java
@@ -26,7 +26,7 @@ public class ConsoleApplication {
     }
 
     public static void main(String[] args){
-        if (args.length == 0 || args[0].equals("help")) {
+        if (args == null || args.length == 0 || args[0].equals("help")) {
             printUsage();
             System.exit(-1);
         }

--- a/src/main/java/Repositories/NPM/NpmIdGenerator.java
+++ b/src/main/java/Repositories/NPM/NpmIdGenerator.java
@@ -12,14 +12,18 @@ import java.util.List;
 import java.util.Properties;
 
 public class NpmIdGenerator implements RepositoryController.IdGenerator {
-
-    private static final String allPackagesUrlGitHub = "https://raw.githubusercontent.com/bconnorwhite/all-package-names/master/data/all.json?raw=true";
+    //private static final String allPackagesUrlGitHub = "https://raw.githubusercontent.com/bconnorwhite/all-package-names/7370fbe7a24c6206cae32c0584cedbd5d2f60616/data/all.json.gz?raw=true";
     private static final String packageArrayJsonKey = "packageNames";
 
     private final Logger logger = LoggerFactory.getLogger(getClass());
 
     private int limit = -1;
     private int offset = 0;
+
+    private static String allPackagesUrlGitHub() {
+        String commitQualifier = System.getProperties().getOrDefault("dgm.npm.commit-qualifier", "master").toString();
+        return "https://raw.githubusercontent.com/bconnorwhite/all-package-names/" + commitQualifier + "/data/all.json.gz?raw=true";
+    }
 
     public NpmIdGenerator() {
         Properties props = System.getProperties();
@@ -33,12 +37,14 @@ public class NpmIdGenerator implements RepositoryController.IdGenerator {
 
         logger.debug("Starting to retrieve NPM ids from GitHub...");
 
-        JSONObject responseObj = HttpUtilities.getContentAsJSON(allPackagesUrlGitHub);
+        JSONObject responseObj = HttpUtilities.getGZIPContent(allPackagesUrlGitHub());
 
         if(responseObj != null && responseObj.has(packageArrayJsonKey)){
             JSONArray packageNames = responseObj.getJSONArray(packageArrayJsonKey);
 
-            List<String> plainPackageNames = new ArrayList<>(packageNames.length());
+            logger.debug("Got a total of " + packageNames.length() + " package names");
+
+            List<String> plainPackageNames = new ArrayList<>(Math.min(packageNames.length(), limit));
 
             for(int i = offset ; i < packageNames.length() ; i++){
                 plainPackageNames.add(packageNames.getString(i));

--- a/src/main/java/Repositories/NPM/NpmIdGenerator.java
+++ b/src/main/java/Repositories/NPM/NpmIdGenerator.java
@@ -12,7 +12,7 @@ import java.util.List;
 import java.util.Properties;
 
 public class NpmIdGenerator implements RepositoryController.IdGenerator {
-    //private static final String allPackagesUrlGitHub = "https://raw.githubusercontent.com/bconnorwhite/all-package-names/7370fbe7a24c6206cae32c0584cedbd5d2f60616/data/all.json.gz?raw=true";
+
     private static final String packageArrayJsonKey = "packageNames";
 
     private final Logger logger = LoggerFactory.getLogger(getClass());

--- a/src/main/java/Utilities/CommandUtilities.java
+++ b/src/main/java/Utilities/CommandUtilities.java
@@ -92,7 +92,7 @@ public final class CommandUtilities {
                     logger.warn("Invalid value for '" + key + "', boolean values expected. Supported: 'true', 'false'");
                     return false;
                 }
-                case "dgm.id-file" -> {
+                case "dgm.id-file", "dgm.npm.commit-qualifier" -> {
                     return !value.isBlank();
                 }
                 default -> {

--- a/src/main/java/Utilities/HttpUtilities.java
+++ b/src/main/java/Utilities/HttpUtilities.java
@@ -5,11 +5,14 @@ import org.json.JSONObject;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.*;
 import java.net.URI;
 import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
 import java.time.Duration;
+import java.util.zip.GZIPInputStream;
 
 /**
  *  Class providing basic HTTP utilities
@@ -49,4 +52,48 @@ public class HttpUtilities {
             return null;
         }
     }
+
+    public static JSONObject getGZIPContent(String url){
+        try {
+            HttpRequest request = HttpRequest.newBuilder()
+                    .timeout(Duration.ofMinutes(TimeoutMinutes))
+                    .uri(new URI(url))
+                    .build();
+
+            HttpResponse<InputStream> response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+
+            if(response.statusCode() == 200){
+
+                byte[] buffer = new byte[128];
+                int len;
+
+                String jsonContent;
+
+                try (ByteArrayOutputStream bos = new ByteArrayOutputStream(); GZIPInputStream gzis = new GZIPInputStream(response.body())) {
+
+                    while ((len = gzis.read(buffer)) != -1) {
+                        bos.write(buffer, 0, len);
+                    }
+
+                    jsonContent = bos.toString(StandardCharsets.UTF_8);
+
+                    return new JSONObject(jsonContent);
+                }
+            } else {
+                throw new RuntimeException("Got non-success status code when downloading zipped data: " + response.statusCode());
+            }
+
+
+
+        } catch(Exception ex){
+            if(GlobalUtilities.isBoolPropertyEnabled("dgm.drop-http-errors"))
+                logger.debug("Exception while downloading JSON: " + ex.getClass() + " -> " + ex.getMessage());
+            else
+                ExceptionLogger.add(ex, HttpUtilities.class.getName());
+
+            return null;
+        }
+    }
+
+
 }

